### DISCUSSION TITLE
Info Request - eigenmode coefficients the same for sources

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#1990] [Q&A]
I'm running simulations to try to determine how long do we need to make a waveguide before a `GaussianBeamSource` couples into the same eigenmode as an `EigenModeSource`. I'm comparing the eigenmode coefficient/power returned by `get_eigenmode_coefficient` at several mode monitors at different locations along a straight waveguide. However, in every simulation I ran I got that the power at each location is the same for both sources.

![modes1](https://user-images.githubusercontent.com/37753156/157534511-9309d48a-b44e-4df1-96c8-651d7c0b6130.jpg)

What causes this? Am I setting something up incorrectly or is this intended?

Also, regardless of which mode I specify for the EigenModeSource (tried band numbers 1,2,5) only mode 2 is excited.

**Comment by smartalecH** [comment]
It's hard to say without any code or even an image of the simulation (easily provided using `plot2D`). There's not a lot of detail about the sources themselves either...

**Reply by Andeloth** [reply]
This is the code for the simulation.

```
import meep as mp
import numpy as np
import matplotlib.pyplot as plt
mp.verbosity(0)

resolution = 30 # pixels per micrometer
Sx = 8
Sy = 10
cell_size = mp.Vector3(Sx,Sy)
pml_width = 1
pml_layers = [mp.PML(thickness=pml_width)]

lambda1 = 1.55
lambda2 = lambda1/2
band_numbers = [i for i in range(1,6)]    # band number (mode index) of eigenmode.

fcen1 = lambda1
df = 0
nfreq = 1

w = 1.0 # width of waveguide
geometry = [mp.Block(center=mp.Vector3(),
                     size=mp.Vector3(w,mp.inf),
                     material=mp.Medium(index=2.1))]

kpoint = mp.Vector3(y=1) # Tells which direction to find the mode in

gaussian_sources = [mp.GaussianBeamSource(mp.ContinuousSource(wavelength=lambda1),
                                          component=mp.Ey,
                                          center=mp.Vector3(0, -Sy/2+2*pml_width),
                                          size=mp.Vector3(w,0),
                                          beam_kdir=kpoint,
                                          beam_w0=w,
                                          beam_E0=mp.Vector3(x=1))]

eigenmode_sources = [mp.EigenModeSource(mp.ContinuousSource(wavelength=lambda1),
                                        center=mp.Vector3(0, -Sy/2+2*pml_width),
                                        size=mp.Vector3(w,0),
                                        direction=mp.Y,
                                        eig_kpoint=kpoint,
                                        eig_band=3,
                                        eig_match_freq=True)]

gauss_sim = mp.Simulation(cell_size=cell_size,
                          resolution=resolution,
                          boundary_layers=pml_layers,
                          sources=gaussian_sources,
                          geometry=geometry)

eig_sim = mp.Simulation(cell_size=cell_size,
                        resolution=resolution,
                        boundary_layers=pml_layers,
                        sources=eigenmode_sources,
                        geometry=geometry)

monitor_positions = np.linspace(-Sy/2 + 2*pml_width+0.1, Sy/2 - pml_width, 10)
monitor_vectors = [mp.Vector3(y=mon_pos) for mon_pos in monitor_positions]
mode_regions = [mp.FluxRegion(center=mon_vec, size=mp.Vector3(x=w))
                for mon_vec in monitor_vectors]  # Center Coordinate of the Eigenmonitor
eig_monitors_gauss_sim = []
eig_monitors_eigen_sim = []
for mode_reg in mode_regions:
    eig_monitors_gauss_sim.append(gauss_sim.add_mode_monitor(fcen1, df, nfreq, mode_reg))
    eig_monitors_eigen_sim.append(eig_sim.add_mode_monitor(fcen1, df, nfreq, mode_reg))

gauss_sim.run(mp.to_appended("out/ex-gauss", mp.at_every(0.5, mp.output_efield_x)), until=100)
eig_sim.run(mp.to_appended("out/ex-eigen", mp.at_every(0.5, mp.output_efield_x)), until=100)

gauss_source_mode_results = [gauss_sim.get_eigenmode_coefficients(monitor, band_numbers) for monitor in eig_monitors_gauss_sim]
eigen_source_mode_results = [eig_sim.get_eigenmode_coefficients(monitor, band_numbers) for monitor in eig_monitors_eigen_sim]
gauss_alpha = np.array([result.alpha[:,0,0] for result in gauss_source_mode_results])
eigen_alpha = np.array([result.alpha[:,0,0] for result in gauss_source_mode_results])
gauss_power = np.abs(gauss_alpha)**2
eigen_power = np.abs(eigen_alpha)**2

plt.figure(figsize=(10,6))
for i in range(gauss_power.shape[-1]):
    plt.plot(monitor_positions - monitor_positions[0], gauss_power[:,i], 'o', label=f'mode {i+1} with gaussian beam source')
    plt.plot(monitor_positions - monitor_positions[0], eigen_power[:,i], label=f'mode {i+1} with eigenmode source')
plt.legend(loc='right')
plt.xlabel(r'y position in simulation ($\mu m$)')
plt.ylabel('power carried by eigenmode')
plt.yscale('linear')
plt.title('Power carried in eigenmode at different positions in the waveguide')
plt.show()
```

**Reply by Andeloth** [reply]
and an image of the simulation

![image](https://user-images.githubusercontent.com/37753156/157539345-633e3e50-4b45-4fae-a163-dd175d3c7161.png)

**Reply by Andeloth** [reply]
It seems like modes 1 and 2 are the exact same, this might be because the fundamental TE and TM modes have the same properties other than TE/TM-ness. What I'm more confused about is why the GaussianBeamSource and EigenModeSource give such similar results. I would expect them to give different results while the gaussian source hasn't coupled into the waveguide eigenmodes yet.

**Reply by smartalecH** [reply]
Your `ModeSource` needs to span at least some of the cladding. Your `GaussianBeamSource` needs to at least span the FWHM (usually much more). Your mode monitors need to span into the cladding too.

Also, in 2D, a polarized source won't excite the other polarization.

**Reply by Andeloth** [reply]
Okay, I increased all of the widths, and I got these simulations (1st gaussian beam source, 2nd eigenmode source with mode 3)
![ex-gauss-1](https://user-images.githubusercontent.com/37753156/157540902-7db0df2e-d9aa-4637-8cc3-697f3369b213.gif)

![ex-eigen-1](https://user-images.githubusercontent.com/37753156/157540925-50721c97-b041-414b-83b2-1957b0d9326e.gif)

and yet the eigenmode coefficients were still only relevant for modes 1 & 2 (which are overlapping)

![image](https://user-images.githubusercontent.com/37753156/157541010-5279bdc1-e87b-4a72-b837-8e655088a9fe.png)

and the GaussianBeamSource and EigenModeSource simulations still give the exact same results.

This is what the simulation looks like now:
![image](https://user-images.githubusercontent.com/37753156/157541074-c7e29748-f2c7-40db-8bfb-00d42d5dc84c.png)

**Reply by Andeloth** [reply]
OH nevermind, I found a mistake in the code that caused that, it's giving me the correct results now.

![image](https://user-images.githubusercontent.com/37753156/157541456-2d3a6dbd-e42d-4cce-8ce5-d1db42fbdaac.png)
